### ISSUE TITLE
Amplification of connection migration attacks

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#1451]
An [NDSS symposium paper by Yuri Gbur and Florian Tschorsch](https://www.ndss-symposium.org/wp-content/uploads/2023/02/ndss2023_s72_paper.pdf) shows that picoquic can amplify a connection migration attack. In this attack, an attacker establishes a regular QUIC connection with a server, and then uses secrets acquired when setting up that connection to send forged packets to the server. These packets could be of two forms:

* NAT variant: simulate a NAT traversal by sending regular 1-RTT packets using the current CID but replacing the source address by the address of a target;
* Challenge variant: simulate a connection migration by sending a packet containing a "source challenge" with a new CID from the spoofed source address of a target.

The focus of the NDSS paper was the "request forgery" attack in which a client tricks a server to send packets to a third party, running a different application. The connection migration attacks are not very powerful there, because only up to 20 bytes of the packet sent by the server can be predicted by the attacker -- and the server will normally refuse to send such packets to "suspicious addresses" per PR #1450. But the method can still be used to generate "volumetric" attacks against a selected target.

In the challenge variant, the current picoquic server will respond with a challenge of its own, in a packet padded to the minimum required 1200 bytes. As the attacker will not acknowledge this packet, picoquic will repeat it a couple of times. We have an amplification factor 3: the attacker sends a 1200 bytes challenge, picoquic repeats the challenge + response packet three times. In theory, this is not very interesting for the attacker, because it also had to send a couple of full size packets to set up the connection, and then a couple of short packets to keep it alive. However, nothing prevents the attacker from creating many challenges on a single connection.

In the NAT variant, the server that detects a packet that seems affected by NAT rebinding is supposed to send a full size challenge to the "natted" address, and also a parallel challenge to the original address. If the challenge to the original address succeeds, the NAT traversal detection is considered spurious and the server uses the old address. If neither challenge succeeds, the challenge will be repeated three times. The packet triggering all that could be very short, so that's a lot of amplification. The attacker would reply to the third repeat of the "original source" challenge to keep the connection alive, and could then repeat the process. This could turn in a reasonably powerful attack.

The first priority is thus to fix the NAT variant. RFC 9000 is somewhat ambiguous about the size of challenge frames. The text says "An endpoint MUST expand datagrams that contain a PATH_CHALLENGE frame to at least the smallest allowed maximum datagram size of 1200 bytes, unless the anti-amplification limit for the path does not permit sending a datagram of this size." These authorizes nodes to send shorter challenges to avoid amplification, with the understanding that they will test the path MTU later. Picoquic should take advantage of that and send short challenges when NAT traversal is triggered by short packets.

The second priority is to fix amplification. For example, the server could decide to only repeat its own challenges if it receives sufficient data from the peer, keeping the amplification level to just 1. However, this is dangerous. If the server sends a challenge response in the same packet as the server challenge, the peer will consider that it's own challenge is validated. If the packet containing the peer's response gets lost, the server will not repeat its own challenge and the path establishment may fail. A symmetric issue may arise if the server refrains from sending the response immediately: the response from the server might arrive too late, or may be lost if the packet is lost. repeating lost responses could help, but is not perfect.

**Comment by huitema** [comment]
The proposed solution for path challenges is:
```
(client)                           (server)
     --- Challenge -->
        <--- Challenge + response ---
     --- response --->
        <--- Ack (response packet) ---
```
Various packet losses are treated as follow:
* If the client's challenge is lost: the client repeats its challenge.
* If the server's challenge + response is lost: the *client* repeats its challenge. (The server does not, because of amplification.)
* If the client's response is lost: the server does not ack the response.
There is an interop issue: if the client does not receive an ACK for the response packet, will it repeat the response? Or should we rely on something else?

**Comment by huitema** [comment]
In the NAT variant, this is harder. The exchange is:
```
(client)                           (server)
     --- natted packet -->
        <--- Challenge + ack (natted) ---
     --- response --->
           <--- Ack (response packet) ---
 ```
If the challenge packet is lost, the natted packet will have to be repeated. This could trigger a repeat of the challenge, which is OK since the client has been sending more bytes. Same if the server's "challenge +ACK" packet is lost. But will the client repeat a response packet if it is lost? What would that trigger an issue?

**Comment by huitema** [comment]
Another part of amplification comes from repeating the challenge responses sent by the server. This explains why the NDSS paper finds an amplification factor of 3.73. When the path is not validated, the server should not repeat the challenges, but it should also not repeat the challenge responses frames independently.

**Comment by huitema** [comment]
This issue is fully addressed by PR #1452.
